### PR TITLE
 cpp: Change result constructor to take rvalue reference 

### LINKED ISSUE #285: C++: Allow only moving evmc_result to evmc::result
Change constructor from `result(const evmc_result&)` to `result(evmc_result&&)`. That would indicate that the new `result` object takes the responsibility of the result.

See https://github.com/ethereum/evmc/pull/281#pullrequestreview-237242775.

### PR DESCRIPTION
**Pull request by chfast**
Closes #285
Changelog entry is missing.

### CODE REVIEW

File: include/evmc/evmc.hpp [hunk 1]
```diff
@@ -27,7 +27,17 @@ class result : private evmc_result
...
     /// Converting constructor from raw evmc_result.
-    explicit result(evmc_result const& res) noexcept : evmc_result{res} {}
+    ///
+    /// The constructed object takes ownership and is responsible for releasing
+    /// the provided result. Therefore, the rvalue reference is used to
+    /// indicate the ownership transfer and that
+    /// the raw evmc_result object MUST NOT be used after.
+    ///
+    /// @param raw_result  The rvalue reference to the evmc_result object.
+    result(evmc_result&& raw_result) noexcept : evmc_result{raw_result}  // NOLINT
```
Comment (axic) [inline include/evmc/evmc.hpp]: Isn't this in conflict with the move constructor below now?

Comment (chfast) [reply include/evmc/evmc.hpp]: No because this is not the move constructor per definition.

Comment (axic) [reply include/evmc/evmc.hpp]: Ah it was late I suppose and confused `evmc_result` with `result`.

File: include/evmc/evmc.hpp [hunk 2]
```diff
@@ -27,7 +27,17 @@ class result : private evmc_result
     using evmc_result::status_code;
 
     /// Converting constructor from raw evmc_result.
-    explicit result(evmc_result const& res) noexcept : evmc_result{res} {}
```
Comment (axic) [inline include/evmc/evmc.hpp]: Do we consider this as a breaking change?

Comment (chfast) [reply include/evmc/evmc.hpp]: It is. Can be moved to 6.3 but as we are fixing C++ API in 6.2 anyway I think we can also include this change.

**Comment by axic** [comment]
clang fails:
```
/home/builder/project/examples/example_host.cpp:105:16: error: no viable conversion from returned value of type 'evmc_result' to function return type 'evmc::result'
        return res;
               ^~~
```

File: include/evmc/evmc.hpp (hunk 2 above)
Comment (axic) [reply include/evmc/evmc.hpp]: I don't think this is fixing a bug in 6.2.x, but improving the API. Following that logic it should be added to 6.3.x, if I'm completely pragmatic.

**Comment by chfast** [comment]
Closing this because I don't like it and in `std::unique_ptr` you can use the pattern we want to prevent here:
```
auto raw = new X{};
auto p = std::unique_ptr{raw};
// Do not: delete raw;
```

**Pull request closed without merging**